### ISSUE TITLE
tensor values between dali and pytorch dataloaders are exactly the same but model performance dropped more than 1%

### ISSUE THREAD

**Issue by tian-yi-zefr** [Lightning-AI/pytorch-lightning#13066]
hey guys, I found one weird behavior in DDP training when I switched pytorch dataloader to dali dataloader, not sure if it's dali/pytorch/pytorch lightning issue. I checked that tensor values between pytorch dataloader and dali dataloader (by `check_dataloader_tensor_values`) when shuffle is off are exactly the same, but the model trained with dali dataloader consistently ends up with 1-2% higher validation loss while their config are exactly the same. This behavior is quite consistent, in my 5 times full training, dali dataloader always produced 1-2% higher validation loss. I made a reproducible example for this behavior and wonder what may cause the validation loss drop or how to make the validation loss closer? Appreciate any hints/suggestions/thoughts! 

sample log from weights and bias 
```
dali dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.26073
	wandb: train_epoch_sample_size 60000.0
	wandb:         train_step_loss 0.27232
	wandb:     trainer/global_step 1179
	wandb:   val_epoch_sample_size 10000.0
	wandb:                val_loss 0.27608
	wandb:           val_step_loss 0.27608

native pytorch dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.25656
	wandb: train_epoch_sample_size 60000.0
	wandb:         train_step_loss 0.1342
	wandb:     trainer/global_step 1179
	wandb:   val_epoch_sample_size 10000.0
	wandb:                val_loss 0.27474
	wandb:           val_step_loss 0.27474
```

# dependency list
```
torch==1.9.0
pytorch-lightning==1.5.10
torchvision=0.10.0
tfrecord=1.14.1
nvidia-dali-cuda110=1.12.0
```

# reproducible example 
```python
import warnings
warnings.filterwarnings("ignore")

import torch
from torch.nn import functional as F
from torch import nn
from pytorch_lightning.core.lightning import LightningModule
from pytorch_lightning import Trainer
from torch.optim import Adam
from torchvision.datasets import MNIST
from torchvision import transforms
from torch.utils.data import DataLoader

import tfrecord 
import numpy as np 
from multiprocessing import Pool, cpu_count
import subprocess
import os

import urllib
opener = urllib.request.build_opener()
opener.addheaders = [('User-agent', 'Mozilla/5.0')]
urllib.request.install_opener(opener)


from pytorch_lightning.utilities.seed import seed_everything
seed_everything(123)


DATA_DIR = './temp'

if not os.path.exists(DATA_DIR):
    os.makedirs(DATA_DIR)

N_GPUS = torch.cuda.device_count() 
print (f"N_GPUS = {N_GPUS}")
# N_CPUS = cpu_count()
N_CPUS = 8
BATCH_SIZE = 64


# load all mnist data
transform = transforms.Compose([transforms.ToTensor()])

# define native pytorch dataloader 
_mnist_train = MNIST(os.getcwd(), train=True, download=True, transform=transform)
native_torch_train_dataloader = DataLoader(_mnist_train, batch_size=BATCH_SIZE, 
    shuffle=False, num_workers=N_CPUS, pin_memory=True)
_mnist_valid = MNIST(os.getcwd(), train=False, download=True, transform=transform)
native_torch_valid_dataloader = DataLoader(_mnist_valid, batch_size=BATCH_SIZE,
    shuffle=False, num_workers=N_CPUS, pin_memory=True)


# write tfrecords  
def _convert_to_1d_list(x):
    if isinstance(x, np.ndarray):
        return x.flatten().tolist()
    elif isinstance(x, list):
        return np.array(x).flatten().tolist()
    else:
        return [x]


def convert_array_to_tfrecod(
    feature_array: np.ndarray, 
    label_array: np.ndarray, 
    file_index: int, 
    file_pattern: str, 
    index_pattern: str
):
    tfrecord_path = file_pattern.format(file_index)
    index_path = index_pattern.format(file_index)
    writer = tfrecord.TFRecordWriter(tfrecord_path)
    
    for f,l in zip(feature_array, label_array): 
        # write tfrecord files by tfrecord library 
        row_record = {"image": (_convert_to_1d_list(f), "float"),
                     "label": (_convert_to_1d_list(l), "int")}
        writer.write(row_record)

    writer.close()
    subprocess.call(["python3", "-m", "tfrecord.tools.tfrecord2idx", tfrecord_path, index_path.format(file_index)])


def convert_array_to_tfrecod_by_multiprocess(
    feature_array: np.ndarray, label_array: np.ndarray, 
    base_path: str, num_workers: int
):
    if not os.path.exists(base_path):
        os.makedirs(base_path)
    
    tfrecord_pattern = os.path.join(base_path, "data_partition_{}.tfrecord")
    index_pattern = os.path.join(base_path, "data_index_{}.index")

    args_ls = [[feature_partition, label_partition, i, tfrecord_pattern, index_pattern] 
        for i,(feature_partition, label_partition) in 
        enumerate(zip(np.array_split(feature_array, num_workers), 
                        np.array_split(label_array, num_workers)))]
    pool = Pool(processes=num_workers) 
    pool.starmap(convert_array_to_tfrecod, args_ls)


def write_mock_data(dl, split):
    image_tensor_ls, label_tensor_ls = [], [] 
    for i,d in enumerate(dl):
        image_tensor_ls.append(d[0])
        label_tensor_ls.append(d[1])
    image_tensor = torch.vstack(image_tensor_ls).numpy()
    label_tensor = torch.cat(label_tensor_ls).numpy()

    convert_array_to_tfrecod_by_multiprocess(image_tensor, label_tensor, 
        os.path.join(DATA_DIR,split), 8)


"""
# write tfrecord files for training loss comparison
for split, dl in [["train", native_torch_train_dataloader], ["valid", native_torch_valid_dataloader]]:
    write_mock_data(dl, split)
"""


########################    DALI related pipeline    ########################
from nvidia.dali import Pipeline
import nvidia.dali.tfrecord as tfrec
from nvidia.dali.plugin.pytorch import DALIGenericIterator
from nvidia.dali.plugin.pytorch import LastBatchPolicy

import nvidia.dali.fn as fn
from nvidia.dali.fn import readers

import glob 
import subprocess
import os 


class TFRecordPipeline(Pipeline):
    def __init__(self, shard_id, num_shards, split, batch_size=BATCH_SIZE, num_threads=N_CPUS, device_id=None):
        super(TFRecordPipeline, self).__init__(batch_size, num_threads, device_id)
        self.input = readers.tfrecord(
            path=sorted(glob.glob(f'{DATA_DIR}/{split}/*.tfrecord')),
            index_path=sorted(glob.glob(f'{DATA_DIR}/{split}/*.index')),
            features={
                "image": tfrec.VarLenFeature(tfrec.float32, 0.0),
                "label": tfrec.VarLenFeature(tfrec.int64, 0)},
            name='tfrecord_reader',
            num_shards=num_shards,
            shard_id=shard_id,
            pad_last_batch=False, 
            random_shuffle=False, 
            seed=-1,
            stick_to_shard=True  
        )

    def define_graph(self):
        return (fn.reshape(self.input["image"], [-1,28,28]), fn.squeeze(self.input["label"], axes=0)) 


class LightningWrapper(DALIGenericIterator):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

    def __next__(self):
        out = super().__next__()
        out = out[0]
        return [out["image"], out["label"]]


def make_dali_dataloader(shard_id, num_shards, split, output_map=["image", "label"]):
    pipe = TFRecordPipeline(shard_id, num_shards, split)
    dali_iter = LightningWrapper(
        pipe, output_map=output_map, 
        reader_name='tfrecord_reader', auto_reset=True, 
        last_batch_padded=False, last_batch_policy=LastBatchPolicy.PARTIAL
        )
    
    return dali_iter


def check_dataloader_tensor_values(dl1, dl2):
    assert len(dl1) == len(dl2)
    for d1, d2 in zip(dl1, dl2):
        for t1, t2 in zip(d1, d2):
            assert torch.equal(t1, t2)
        

class TestMNIST(LightningModule):

    def __init__(self, use_dali):
        super().__init__()

        self.use_dali = use_dali
        # mnist images are (1, 28, 28) (channels, width, height)
        self.seq = nn.ModuleList([torch.nn.Linear(28 * 28, 128),
                                  torch.nn.Linear(128, 256),
                                  torch.nn.Linear(256, 10)])

    def forward(self, x):
        batch_size, channels, width, height = x.size()        
        # (b, 1, 28, 28) -> (b, 1*28*28)
        x = x.view(batch_size, -1)
        for i, l in enumerate(self.seq):
            x = self.seq[i](x)

        x = F.log_softmax(x, dim=1)
        return x

    def training_step(self, batch, batch_idx):
        x, y = batch 
        model_out = self.forward(x)
        train_loss = self.cross_entropy_loss(model_out, y)
        self.log("train_step_loss", train_loss)
        output = {"loss": train_loss, "train_batch_size": model_out.shape[0]}
        return output
    
    def training_epoch_end(self, outputs):
        outputs = self.all_gather(outputs)
        train_epoch_sample_size = torch.stack([x["train_batch_size"] for x in outputs]).sum().type(torch.FloatTensor)
        train_loss_mean = torch.stack([x["loss"]*x["train_batch_size"] for x in outputs]).sum()/train_epoch_sample_size
        self.log("train_epoch_loss", train_loss_mean, prog_bar=False, on_epoch=True)
        self.log("train_epoch_sample_size", train_epoch_sample_size, prog_bar=False, on_epoch=True)  

    def validation_step(self, batch, batch_idx):
        x, y = batch
        model_out = self.forward(x)
        val_loss = self.cross_entropy_loss(model_out, y)
        self.log("val_step_loss", val_loss) 
        output = {"loss": val_loss, "val_batch_size": model_out.shape[0]}
        return output

    def validation_epoch_end(self, outputs):
        outputs = self.all_gather(outputs)
        val_epoch_sample_size = torch.stack([x["val_batch_size"] for x in outputs]).sum().type(torch.FloatTensor)
        val_loss_mean = torch.stack([x["loss"]*x["val_batch_size"] for x in outputs]).sum()/val_epoch_sample_size
        self.log("val_loss", val_loss_mean, prog_bar=False, on_epoch=True)  
        self.log("val_epoch_sample_size", val_epoch_sample_size, prog_bar=False, on_epoch=True)  
       
    def cross_entropy_loss(self, logits, labels):
        return F.nll_loss(logits, labels)

    def configure_optimizers(self):
        return Adam(self.parameters(), lr=1e-3)

    def train_dataloader(self):
        if self.use_dali:
            return make_dali_dataloader(shard_id=self.global_rank, num_shards=self.trainer.world_size, split="train")
        else:
            return native_torch_train_dataloader
    
    def val_dataloader(self):
        if self.use_dali:
            return make_dali_dataloader(shard_id=self.global_rank, num_shards=self.trainer.world_size, split="valid")
        else:
            return native_torch_valid_dataloader


# check values are the same
_dali_train_dl = make_dali_dataloader(shard_id=0, num_shards=1, split="train")
_dali_valid_dl = make_dali_dataloader(shard_id=0, num_shards=1, split="valid")
check_dataloader_tensor_values(_dali_train_dl, native_torch_train_dataloader)
check_dataloader_tensor_values(_dali_valid_dl, native_torch_valid_dataloader)


from pytorch_lightning.loggers import WandbLogger

# True for dali dataloader run and False for native pytorch dataloader run 
USE_DALI = True 
trainer_logger = WandbLogger(project=f"use_dali_dataloader_is_{USE_DALI}")

# train with multiple GPUs
model = TestMNIST(use_dali=USE_DALI)
trainer = Trainer(
    gpus=torch.cuda.device_count(), 
    strategy='ddp',
    min_epochs=10,
    max_epochs=10,
    deterministic=True,
    logger=trainer_logger
)
trainer.fit(model)

```

cc @awaelchli @rohitgr7 @akihironitta @justusschock @ninginthecloud @otaj @kaushikb11

**Comment by tianyi-qi-zefr** [comment]
> not

# after setting replace_sampler_ddp=False
```
dali dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.26073
	wandb: train_epoch_sample_size 60000.0
	wandb:         train_step_loss 0.27232
	wandb:     trainer/global_step 1179
	wandb:   val_epoch_sample_size 10000.0
	wandb:                val_loss 0.27608
	wandb:           val_step_loss 0.27608

native pytorch dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.28332
	wandb: train_epoch_sample_size 480000.0  # sample size is wrong, it should be 60k
	wandb:         train_step_loss 0.33183
	wandb:     trainer/global_step 9379
	wandb:   val_epoch_sample_size 80000.0
	wandb:                val_loss 0.3326
	wandb:           val_step_loss 0.3326
```

# after setting replace_sampler_ddp=False and gpus=1
```
dali dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.283
	wandb: train_epoch_sample_size 60000.0
	wandb:         train_step_loss 0.08335
	wandb:     trainer/global_step 9379
	wandb:   val_epoch_sample_size 10000.0
	wandb:                val_loss 0.32135
	wandb:           val_step_loss 0.32135

native pytorch dataloader run 
	wandb:                   epoch 9
	wandb:        train_epoch_loss 0.28332
	wandb: train_epoch_sample_size 60000.0
	wandb:         train_step_loss 0.33183
	wandb:     trainer/global_step 9379
	wandb:   val_epoch_sample_size 10000.0
	wandb:                val_loss 0.3326
	wandb:           val_step_loss 0.3326
```

when replace_sampler_ddp=False, the sample size on 8 GPU is wrong, it should be 60k but got 480k. with replace_sampler_ddp=False and gpus=1, the validation percentage is even bigger.

**Comment by justusschock** [comment]
That the sample size is wrong is expected in that case, as each GPU gets 60k samples (all of them) and 8*60k=480k

However, that the loss goes up on one gpu is certainly something I do not have an explanation for. That being said, I don't really know a lot about dali, so can't tell much what they are doing under the hood. It seems like they are doing something similar under the hood by shard ids. Maybe it really is due to which/how samples are repeated if necessary.

**Comment by tianyi-qi-zefr** [comment]
@justusschock the dali dataloader reads sample by shard id, I verified that dali dataloader don't repeat sample in multi GPU env. I also tried to drop last paritial batch in both dataloaders, the validation loss change is still 0.27159 vs 0.28254 in pl==1.6.3.
